### PR TITLE
[MetaSchedule] Refactor ScheduleRule Attributes

### PR DESCRIPTION
**Pull request by junrushao**
This PR refactors the existing `schedule_rule` block annotation-based dispatch into a more organized folder structure, which follows the exact same mechanism as AutoTVM.

In the example below, when target is `Target("llvm -keys=cuda,gpu")` and the block annotation is `conv2d_nchw_winograd_inverse`:

```python
with T.block("some_block"):
  T.block_attr({
    "scehdule_rule": "conv2d_nchw_winograd_inverse",
  })
```

the system will find the following global packed functions in order:
- `meta_schedule.cuda.conv2d_nchw_winograd_inverse`
- `meta_schedule.gpu.conv2d_nchw_winograd_inverse`
whose function signatures are:

```python
def schedule_rule(
  sch: tvm.tir.Schedule,
  block: tvm.tir.schedule.BlockRV,
) -> List[tvm.tir.Schedule]:
```

In terms of code organization, for example, for target key `cuda`, the schedule functions are supposed to be defined in:
- `include/tvm/meta_schedule/schedule/cuda` for public methods
- `src/meta_schedule/schedule/cuda` for private methods
- `python/tvm/meta_schedule/schedule/cuda` for direct python side definition

### CODE REVIEW

**Review by spectrometerHBH — APPROVED** [review]

File: python/tvm/relay/op/strategy/cuda.py
Comment (junrushao) [reply python/tvm/relay/op/strategy/cuda.py:170]: certainly - it definitely sounds better

File: python/tvm/topi/nn/conv2d.py
Comment (junrushao) [reply python/tvm/topi/nn/conv2d.py:1090]: not for simplicity but for consistency with auto-scheduler's nhwc case a couple of lines above. note that this configuration is also overridden by cuda dispatch, so we dont have to worry too much about it

File: python/tvm/meta_schedule/search_strategy/evolutionary_search.py
Comment (masahi) [inline python/tvm/meta_schedule/search_strategy/evolutionary_search.py:61]: cc @zxybazh This change so small and looks unrelated to this PR, but it has huge implications. Was there enough discussion to make this change? Tuning time will become much shorter, which I like, but could there be a concern for perf regression due to this change?

Comment (junrushao) [reply python/tvm/meta_schedule/search_strategy/evolutionary_search.py:61]: @masahi Ask for asking, and it's definitely a valid concern. The reason is that we found it's somehow set to an unreasonably large number (we were consistently using 512 on downstream), and now wanted to scale it back. Also note that the population calculation is slightly different than AutoScheduler, which includes invalid candidates in population, while ours always prune ahead of time.

We did make sure we have proper numbers in hand before merging, and please refer to the table below for details:

|                     | Mainline (ms) | This PR (ms) | Difference |
|---------------------|---------------|--------------|------------|
| resnet_50      |   1.829128008 |   1.73111331 |      5.66% |
| mobilenet_v2   |  0.4773168361 | 0.4806015715 |     -0.68% |
| resnet_18      |   0.678600832 | 0.6301909627 |      7.68% |
| mobilenet_v3   |  0.6649458484 |  0.668666894 |     -0.56% |
| wide_resnet_50 |   3.753781691 |  3.077549256 |     21.97% |
| densenet_121   |   2.361973117 |  2.288411393 |      3.21% |
| inception_v3   |    3.53431975 |  3.478368823 |      1.61% |
| resnet3d_18    |   7.791658449 |  7.465777971 |      4.36% |

File: src/meta_schedule/schedule_rule/apply_custom_rule.cc
Comment (masahi) [inline src/meta_schedule/schedule_rule/apply_custom_rule.cc:88]: Is this a typo? Shouldn't it be `ScheduleRuleApplyCustomRule`.

`schedule_rule/apply_custom_rule.py` refers to `ScheduleRuleApplyCustomRule` that doesn't exist.

File: src/meta_schedule/space_generator/post_order_apply.cc
Comment (masahi) [inline src/meta_schedule/space_generator/post_order_apply.cc:146]: It seems this change broke auto tensorization for VNNI and Hexagon `vrmpy`. They target the TE compute that happens to be annotated with `schedule_rule` (specifically, NCHWc int8 conv2d https://github.com/apache/tvm/blob/main/python/tvm/topi/nn/conv2d.py#L534-L552). After this change, `MultiLevelTilingWithIntrin` will never be applied to the block corresponding to this te compute.

So for auto tensorization, we just want to ignore this annotation.

File: src/meta_schedule/schedule_rule/apply_custom_rule.cc
Comment (zxybazh) [reply src/meta_schedule/schedule_rule/apply_custom_rule.cc:88]: I think this should be a typo, given it's part of a `ScheduleRule` and the ffi function call [here](https://github.com/apache/tvm/blob/b16a64d6edb9fd1a014fc51995dff7d0e2f4c84e/python/tvm/meta_schedule/schedule_rule/apply_custom_rule.py#L32) on the python side doesn't have corresponding registry.

Comment (junrushao) [reply src/meta_schedule/schedule_rule/apply_custom_rule.cc:88]: Yep this is a typo. Feel free to send a quick patch :-)

File: src/meta_schedule/space_generator/post_order_apply.cc
Comment (junrushao) [reply src/meta_schedule/space_generator/post_order_apply.cc:146]: Ah I see! I was not aware the usecase where schedule_rule is hardcoded but **intentionally** ignored. In this particular case, how about we add a line below in `ApplyCustomRule`:

```C++
sch->Unannotate(block, "schedule_rule");
```

File: src/meta_schedule/schedule_rule/apply_custom_rule.cc
Comment (zxybazh) [reply src/meta_schedule/schedule_rule/apply_custom_rule.cc:88]: Just sent a patch #13353

File: src/meta_schedule/space_generator/post_order_apply.cc
Comment (masahi) [reply src/meta_schedule/space_generator/post_order_apply.cc:146]: Thanks `Unannotate` fixed the issue.

File: src/te/operation/create_primfunc.cc
Comment (masahi) [inline src/te/operation/create_primfunc.cc:120]: Is this change intentional? Just found that this change broke `RewriteLayout` when `link-params = True`.

Comment (junrushao) [reply src/te/operation/create_primfunc.cc:120]: The problem is that the Buffer objects in the annotation will not be updated when doing transformation, and that’s why I removed this line. I was not aware this is useful for link-param-specific hack.

Feel free to add it back if it’s needed. Shall we add a flag in pass config which defaults to off, and hexagon-specific users could turn it on when needed?

Comment (masahi) [reply src/te/operation/create_primfunc.cc:120]: I see, for now I reverted this change in https://github.com/apache/tvm/pull/13384 to unblock tuning with `link params`. 

I'd say the same problem applies to the current approach which specifies the layout free buffers by their indices in the parameter list, since it is also technically possible (although unlikely) that some transformation may change the signature of the prim func, which invalidates the parameter indices.

**Pull request merged**
